### PR TITLE
[framework] merge guid into account

### PR DESCRIPTION
**Pull request by davidiw**
This reduces storage explosion on new accounts by 33%.
Unfortunately there's a huge amount of quirks:
* created aptos_account as an entity that knows about everything
* moved coin out of account and into aptos_account to avoid cyclical
  dependencies
* removed coins, since we now are going a different route with
  dependencies
* moved guid and events to framework as they are now directly dependent
  on account
* moved prologue and epilogue out of account since it knows about coin

### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3451)

### CODE REVIEW

File: aptos-move/aptos-vm/src/errors.rs [hunk 1]
```diff
@@ -66,6 +67,7 @@ pub fn convert_prologue_error(
             VMStatus::Error(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION)
         }
         VMStatus::MoveAbort(location, code) => {
+            println!("{:?} {:?}", location, code);
```
Comment (gregnazario) [inline aptos-move/aptos-vm/src/errors.rs:70]: Should remove this println

File: aptos-move/aptos-vm/src/errors.rs [hunk 2]
```diff
@@ -46,15 +46,16 @@ fn error_split(code: u64) -> (u8, u64) {
 /// Any non-abort non-execution code is considered an invariant violation, specifically
 /// `UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION`
 pub fn convert_prologue_error(
-    chain_specific_info: &ChainSpecificAccountInfo,
+    transaction_validation: &TransactionValidation,
     error: VMError,
     log_context: &AdapterLogSchema,
 ) -> Result<(), VMStatus> {
     let status = error.into_vm_status();
+    println!("{:?}", status);
```
Comment (gregnazario) [inline aptos-move/aptos-vm/src/errors.rs:54]: Same, please remove

**Review by gregnazario — COMMENTED** [review]
This is a pretty significant change, I'll let someone else review the move code

**Comment by gregnazario** [comment]
I'm concerned that any app built on top of events, accounts, or coins (which is all of them) has a pretty significant breaking change

File: crates/aptos/src/move_tool/mod.rs [hunk 3]
```diff
@@ -137,7 +137,8 @@ impl CliCommand<()> for InitPackage {
         init_move_dir(
             package_dir.as_path(),
             &self.name,
-            Some("devnet".to_string()),
+            // TODO(@davidiw) return back to main once lands and then cherry-pick for devnet
+            Some("conversions".to_string()),
             addresses,
             self.prompt_options,
         )
```
[the remark this answers is not in the document]
Comment (davidiw) [reply crates/aptos/src/move_tool/mod.rs:144]: Sadly :(

**Comment by davidiw** [comment]
Last night we came to the realization that we were going to create 3 resources per account. This reduces it to two. Fortunately this isn't as crazy a breaking change as I would have expected. It certainly breaks the current account and anyone that might read the generator (did anyone except our tests?). This also cleans up some of the coin::register pains that we'd gotten a lot of feedback about.

In terms of a breaking change to external parties, the test case will definitely be pained by having to create accounts, but the actual internal app just swaps from event::* / guid::* to account::*

File: aptos-move/framework/aptos-framework/sources/guid.move [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        /// If creation_num is `i`, this is the `i+1`th GUID created by `addr`
+        creation_num: u64,
+        /// Address that created the GUID
+        addr: address
+    }
+
+    /// A capability to create a privileged identifier on behalf of the given address
+    struct CreateCapability has key, store, drop {
+        addr: address
+    }
```
Comment (msmouse) [inline aptos-move/framework/aptos-framework/sources/guid.move:21]: remove?

File: aptos-move/framework/aptos-framework/sources/aptos_account.move [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    ///
+    const ENO_VALID_FRAMEWORK_RESERVED_ADDRESS: u64 = 11;
+
+    /// Prologue errors. These are separated out from the other errors in this
+    /// module since they are mapped separately to major VM statuses, and are
+    /// important to the semantics of the system.
+    const PROLOGUE_EINVALID_ACCOUNT_AUTH_KEY: u64 = 1001;
+    const PROLOGUE_ESEQUENCE_NUMBER_TOO_OLD: u64 = 1002;
+    const PROLOGUE_ESEQUENCE_NUMBER_TOO_NEW: u64 = 1003;
+    const PROLOGUE_EACCOUNT_DOES_NOT_EXIST: u64 = 1004;
```
Comment (msmouse) [inline aptos-move/framework/aptos-framework/sources/aptos_account.move:40]: some error codes not referenced

**Comment by msmouse** [comment]
> Also if this is just about the resource scarcity, we should discuss if we not better solve this problem by grouping of access paths and let the storage system deal with it. 

@wrwg Not sure if we are completely on the same page, but the grouping of access paths must by explicit in the language so the developers (of the framework and other smart contracts) has full control and in the end things that are accessed together will be stored together, and provided our gas schedule one can make conscious trade-off in terms of the granularity of storage.

That said, I think this PR expressed that concept "guid is actually part of the account, and access control is though the account"  with the existing language structures (structs composition). I don't see it compromise much in the space of cleanness, one would even argue this is cleaner, if you ask @dariorussi  he might say this is perfectly what we should have done a few months ago.

**Comment by wrwg** [comment]
> > Also if this is just about the resource scarcity, we should discuss if we not better solve this problem by grouping of access paths and let the storage system deal with it.
> 
> @wrwg Not sure if we are completely on the same page, but the grouping of access paths must by explicit in the language so the developers (of the framework and other smart contracts) has full control and in the end things that are accessed together will be stored together, and provided our gas schedule one can make conscious trade-off in terms of the granularity of storage.
> 
> That said, I think this PR expressed that concept "guid is actually part of the account, and access control is though the account" with the existing language structures (structs composition). I don't see it compromise much in the space of cleanness, one would even argue this is cleaner, if you ask @dariorussi he might say this is perfectly what we should have done a few months ago.

There is many things which could be 'cleaner'. Pulling this particular one out and making a big breaking change after code freeze doesn't make sense in my mind.

**Comment by sherry-x** [comment]
> Pulling this particular one out and making a big breaking change after code freeze doesn't make sense in my mind.

@wrwg  Unfortunately we only realized this issue yesterday so we're delaying another day for the branch cut to incorporate this change, and hopefully finally after this one we can do the real freeze!

**Review by zekun000 — APPROVED** [review]

**Review by movekevin — APPROVED** [review]
This is not as bad as I expected :)

File: aptos-move/framework/aptos-framework/sources/aptos_account.move [hunk 6]
```diff
@@ -0,0 +1,47 @@
+module aptos_framework::aptos_account {
```
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move:1]: Yeah there are some naming improvements we can make for sure

**Pull request merged**